File: Exclusive time of func.py
# // Time Complexity : O(N)
# // Space Complexity : O(N)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No

class Solution:
    def exclusiveTime(self, n, logs):
        time = [0] * n  # result list
        stack = []  # to keep track of idx
        last_time = 0

        for log in logs:
            # split the strings
            idx, strs, t = log.split(':')
            idx, t = int(idx), int(t)

            # keep track of previous time and update time list accordingly
            if strs == "start":
                if stack:
                    time[stack[-1]] += t - last_time
                stack.append(idx)
                last_time = t

            # pop the idx from stack and update the time list, +1 since end time is inclusive
            else:
                time[stack.pop()] += t + 1 - last_time
                last_time = t + 1

        return time

File: ExclusiveTimeOfFunc.java
// Time Complexity : O(No of logs)
// Space Complexity : O(No of unique funcs)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


class Solution {
    public int[] exclusiveTime(int n, List<String> logs) {
        int[] res = new int[n];
        int prev = 0;
        int curr = 0;
        Stack<Integer> st = new Stack<>();
        for(String log: logs) {
            String[] strArr = log.split(":");
            curr = Integer.parseInt(strArr[2]);
            // at every start of a func,
            // we should pause the currently executing func
            // calculate its start substracting with last time stamp
            if(strArr[1].equals("start")) {
                if(!st.isEmpty()) {
                    res[st.peek()] += curr-prev;
                }
                st.push(Integer.parseInt(strArr[0]));
                prev = curr;
            } else {
                // at everyy end, we neeed to stop the currently executing func
                res[st.pop()] += curr-prev+1;
                prev = curr+1;
            }
        }
        return res;

    }
}